### ISSUE TITLE
origin check for external login servers break users created <cf170

### ISSUE THREAD

**Issue by msmolny** [cloudfoundry/uaa#480]
Mid 2014, with cf170, the column "origin" was introduced. Its default is 'uaa'. See also:

https://github.com/cloudfoundry/uaa/blob/5cbf9c1b65598892e6f72966a1288e11f72e793a/server/src/main/resources/org/cloudfoundry/identity/uaa/db/postgresql/V1_7_0__OriginAndExternalIDColumns.sql

Recent versions of UAA now forbid external login servers (source=login) to use the UAA origin attribute for creating tokens for users. While I understand the idea and concept of this origin attribute, forbidding the usage is a breaking change and invalidates all users created <cf170 or after cf170 if created with origin=uaa - which was allowed until recently.

This check is too restrictive and breaks compatibility, so I suggest to remove this checking.

Story that introduced this: https://www.pivotaltracker.com/n/projects/997278/stories/119860913
Referenced code change: https://github.com/cloudfoundry/uaa/blob/develop/server/src/main/java/org/cloudfoundry/identity/uaa/authentication/manager/LoginAuthenticationManager.java#L130

**Comment by fhanik** [comment]
Am I correctly assuming this is the check we are talking about?

        if(info.get(OriginKeys.ORIGIN)!=null && info.get(OriginKeys.ORIGIN).equals(OriginKeys.UAA)){
            throw new BadCredentialsException("uaa origin not allowed for external login server");
        }

Since the user created has `password=""`, enabling the creation of `origin=uaa` would mean I can bypass the login server, go directly to the UAA token endpoint with a *known* password.

**Comment by fhanik** [comment]
Original story link

https://www.pivotaltracker.com/n/projects/997278/stories/119860913

**Comment by msmolny** [comment]
Yes, this is the line I am talking about. We did some tests and verified that it is not possible to go directly to the UAA and get a UAA token.
Until CF170, the column "origin" even did not existed, so previously, there must have been another indicator than the origin column that this user can only be authenticated by the login server.

Do you agree that this change in the code is breaking backward compatibility to users created prior to CF170 using the login server?

**Comment by fhanik** [comment]
@msmolny 

Yes, I agree it breaks backwards compatibility. I think that's a fact rather than a consensus claim.

`origin=uaa` is our indicator on what type of user this is, and where the authentication for this user belongs. 

Actions like "reset password" suddenly become available to users that should not be able to authenticate on the UAA.

In the use case where the user is disabled in the IDP, they would still have access using UAA's own mechanism. This is troublesome. I looked into the history of this

Prior to the `origin` column, we had in chronological order
1. random password set for user
2. blank `""` password 
3. origin was set to "login-server" for any request coming through that path
4. We then introduced the ability to set `origin` since the login-server itself desired to differentiate between IDPs it was using. 

The intention was never meant to allow the login-server to override that behavior by setting it to "uaa"
It was a bug that was [fixed](https://github.com/cloudfoundry/uaa/commit/7fa4256ae13d599e925ca3e780525f7b05f54d15). 

The easiest way to identify UAA vs non UAA users that both have origin="uaa" is to run a script that checks the bcrypt match for an empty password string `""` and then commence a migration

If there are no non UAA users, the a simple SQL migration to adjust the value of the column.

I'm going to close this issue, as I don't think that this long after we have released the fix it would be wise to revert the fix and by doing so create some of the issues I mentioned above.
